### ISSUE TITLE
Removing libatomic_ops

### ISSUE THREAD

**Issue by mahrud** [Macaulay2/M2#1113]
libatomic_ops is a library atomic operations  (i.e. those which cannot be interrupted by another thread) before C11 and C++14 standards. Currently the first line on its [README.md](https://github.com/ivmai/libatomic_ops/blob/master/README.md) says:
> IN NEW CODE, PLEASE USE C11 OR C++14 STANDARD ATOMICS INSTEAD OF THIS PACKAGE.

This issue is for using either the [C++14 atomic](https://en.cppreference.com/w/cpp/atomic) or the [C11 atomic](https://en.cppreference.com/w/c/atomic) operations. These operations are part of the language, so don't require an extra library to be installed. They work much better with the C++ [thread](https://en.cppreference.com/w/cpp/thread) library and are also much better documented than [libatomic_ops](https://www.hpl.hp.com/research/linux/atomic_ops/README.txt).

libatomic_ops mostly used in atomic.d and system, but the uses propagate from there. I tried refactoring atomic.d to use C11 atomic, but ran into issues because atomic.d is later used in C++, where `_Atomic` type is not defined (and `extern "C"` didn't help). On the other hand, C++14 atomic is not usable in C.

Not sure how to proceed, but moving to C11 & C++14 would definitely make multithreading easier.

**Comment by mahrud** [comment]
On a second look, maybe this isn't impossible and the C and C++ atomic libraries are compatible. Here is a test: 
- c11.c:
```
#include <pthread.h>
#include <stdatomic.h>
#include <stdio.h>

atomic_int acnt;
int cnt;

void adding()
{
  for(int i=0; i<10000; i++)
    {
      acnt++;
      cnt++;
    }
  pthread_exit(NULL);
}
```
- cxx14.cpp:
```
#include <vector>
#include <atomic>
#include <thread>

using namespace std;

extern "C" void adding();
extern "C" atomic_int acnt;
extern "C" int cnt;

int main()
{
  vector<thread> v;

  for(size_t i {0}; i < 10; ++i) {
    v.emplace_back(adding);
  }

  for(auto &t : v) { t.join(); }

  printf("the value of acnt is %d\n", acnt.load());
  printf("the value of cnt is %d\n", cnt);

  return 0;
}
```
We should be able to compile separately and link together like so:
```
  gcc -c c11.c -o c11.o --std=c11
  g++ -c cxx14.cpp -o cxx14.o --std=c++14
  g++ cxx14.o c11.o -pthread
  ./a.out
```
Which should give an output like:
```
the value of acnt is 100000
the value of cnt is 29759
```
Indicating that the atomic counter declared and incremented in `c11.c` is safely accessed from `cxx14.cpp` using atomic operations.

**Comment by DanGrayson** [comment]
This seems to be low priority, since the current code works and is already written.

**Comment by mahrud** [comment]
The threading system isn't really working. M2 starting 16 threads on startup but never uses more than one. First step in improving the system is using libraries that have good documentation so that we can understand what the code is doing. This is directly related to #143, #175, and making sure #1099 is done safely.

**Comment by mahrud** [comment]
@DanGrayson I think this might be a good place to start improving the task supervisor, since atomic operations and threading are closely tied in C/C++. This is a rough sketch of steps involved:
- remove dependency on atomic_ops, replace with std::atomic
  - some changes in d/atomic.d
  - some changes in system
  - some changes in scc1
- rethink the way thread local objects are added to the GC root list by the task supervisor
  - some changes in system/supervisorinterface.h
  - some changes in d
  - some changes in scc1 related to thread local variables
- use C++ threads in the task supervisor instead of manually using pthreads
- optionally, experiment with using a different garbage collector while we're here

**Comment by d-torrance** [comment]
Closed by #3082
